OptionSelector/AnimalTypesTableViewController.m

```diff
@@ -0,0 +1,178 @@
+                           @"Russian Blue",
+                           @"Calico",
+                           @"Siamese",
+                           @"British Shorthair",
+                           @"Persian",
+                           @"Cornish Rex",
+                           @"Scottish Fold"];
+    
+    cat.selection = @" ";
+    
+    
+    
+    
+    
+    // breeds of bug - these are our options to choose from
+    
+    bug.options = @[
+                           @"Butterfly",
+                           @"Ladybug",
+                           @"Caterpillar",
+                           @"Spider",
+                           @"Dragonfly",
+                           @"Cricket"];
+    
+    bug.selection = @" ";
+    
+    
+    
+    // adding these CQCategory instancse - dog, cat, bug - to our array to hold the 'types' of animal
+    
+    [self.array addObject:dog];
+    [self.array addObject:cat];
+    [self.array addObject:bug];
+    
+}
+
+- (void)viewWillAppear:(BOOL)animated{
+    [super viewWillAppear:animated];
+    [self.tableView reloadData];
+}
+- (void)didReceiveMemoryWarning {
+    [super didReceiveMemoryWarning];
+    // Dispose of any resources that can be recreated.
+}
+
+
+- (void)addItemViewController:(AnimalDetailTableViewController *)controller didFinishEnteringItem:(CQCategory *)item{
+
+    [self.array replaceObjectAtIndex:self.index withObject:item];
+    
+}
+
+
+
+
+#pragma mark - Table view data source
+
+- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
+    return 1;
+}
+
+- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
+    return self.array.count;
+}
+
+
+- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
+    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"AnimalTypesIdentifier" forIndexPath:indexPath];
+    
+    CQCategory *object = [self.array objectAtIndex:indexPath.row];
+    NSString *name = object.name;
+    
+    cell.textLabel.text = name;
+    cell.detailTextLabel.text = [object.selection isEqualToString:@" "]?@" ":object.selection;
+    return cell;
+}
+
+
+#pragma mark - Navigation
+
+
+- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
+    
+    NSLog(@"You clicked on an animal with a specific breed.");
+    
+    NSIndexPath *indexPath = [self.tableView indexPathForSelectedRow];
+    
+    
+    AnimalDetailTableViewController *detailViewController = segue.destinationViewController;
+    
+    self.index = indexPath.row;
+    
+    if (indexPath.row == 0) {
+        CQCategory *dog = self.array[indexPath.row];
+        detailViewController.thing = dog;
+    } else if (indexPath.row == 1) {
+        CQCategory *cat = self.array[indexPath.row];
+        detailViewController.thing = cat;
+    } else if (indexPath.row == 2) {
+        CQCategory *bug = self.array[indexPath.row];
+        detailViewController.thing = bug;
+    }
```

Comment:
It looks like you may have overthought this code. Does it really matter _which_ index was selected? After all, the result is the same for each: assign the category object to the `thing` property of the view controller. The only thing that distinguishes them is the name of the variable.

The reason I point this out is because your code can be refactored to be more concise and explicit. As it stands, you would have to write new lines of code _for every new category that you add_. That's a lot of code to manage, if you end up adding 100 new categories.

As an alternative, you could just do this:

```objective-c
CQCategory *category = self.array[indexPath.row];
detailViewController.thing = category;
```

This code works the same as yours, but doesn't require conditionals and doesn't require that a new line of code is written for each new category added, or that you even know the order of the categories.

The lesson here is that you should try (whenever possible) to not create too many restrictions on your code. Put another way, you want to be cautious of **introducing dependencies**. These 9 lines of code are dependent on two things that are very likely to change:

1. There are three (only and always) elements in the array `self.array`
2. The first element is (now and always) something that makes sense to assign to the variable `dog`, the second to `cat`, and the third to `bug`.

If at any point later on in time your code changes in one of these ways (or others that we may not be able to predict), then these lines of code will either break or produce bugs:

- The order of the categories changes
- The number of categories changes (one is added or removed)

So when you're working with data structures, especially mutable ones, you want to avoid writing code that expects them to have a certain static shape or content. This way you get into less trouble in the future.